File: Python/Class.py
# coding: utf-8

class User:
    # クラス変数（Public）
    className = None

    # コンストラクタ
    def __init__(self, name, age):
        # インスタンス変数（Public）
        self.name = name
        self.age = age
        # インスタンス変数（Private）
        self.__address = None

    # デストラクタ
    def __del__(self):
        print("destructor")

    # プロパティ（ゲッター）
    @property
    def address(self):
        return self.__address

    # プロパティ（セッター）
    @address.setter
    def address(self, address):
        self.__address = address

    # インスタンスメソッド
    def dump(self):
        return ("name: %s, age: %d, address: %s"
         % (self.name, self.age, self.address))

    # クラスメソッド
    @classmethod
    def classMethod(cls):
        print("class_method")

    # スタティックメソッド
    @staticmethod
    def staticMethod():
        print("static_method")

# 継承
class SpecialUser (User):
    # オーバーライド
    def dump(self):
        return ("SubClass name: %s, age: %d, address: %s"
         % (self.name, self.age, self.address))


# 生成
user1 = User("Honda", 27)
str1 = user1.dump()
print(str1)

# プロパティへのアクセス
user1.address = "Tokyo"
print(user1.address)

# クラス変数
User.className = "ABC"
print(User.className)

# クラスメソッドとスタティックメソッド
User.classMethod()
User.staticMethod()

# サブクラス
user2 = SpecialUser("Kawasaki", 33)
str2 = user2.dump()
print(str2)

File: Ruby/Class.rb
class User
    # インスタンス変数
    @name = ""
    @age = 0

    # イニシャライザ
    def initialize(name)
      puts 'initialize'
      @name = name
    end

    # インスタンスメソッド
    def compute(num1, num2)
        puts 'User Compute'
        return num1 * num2
    end

    # ゲッター
    def getName
        @name
    end
    def getAge
        @age
    end

    # セッター
    def setName=(name)
        @name = name
    end
    def setAge=(age)
        @age = age
    end

    # クラス変数
    @@classVar = "hoge"

    # クラスメソッド
    class << self
        def classMethod
          puts @@classVar
        end
    end
end

# 継承
class SpecialUser < User
    # オーバーライド
    def compute(num1, num2)
        puts 'SpecialUser Compute'
        super # 自動的に引数が渡される
        #return num1 * num2 * 2
    end
end


user = User.new("Honda")
p user.compute(10, 20)
p user.getName
user.setAge = 30
p user.getAge

User.classMethod

specialUser = SpecialUser.new("Kawasaki")
p specialUser.compute(10, 20)

